### ISSUE TITLE
[explicit-function-return-type] option to ignore return value on functions with return type signatures

### ISSUE THREAD

**Issue by aboyton** [typescript-eslint/typescript-eslint#2839]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": ["error"]
  }
}
```

```TS
interface SomeObject {
  hello: string;
  callback: () => string;
}

function foo(): SomeObject {
  return {
    hello: 'world',
    callback: () => 'this is a callback', // Missing return type on function.eslint@typescript-eslint/explicit-function-return-type
  };
}

function bar(): SomeObject {
  return {
    hello: 'world',
    callback: () => 'this is a callback', // This one is fine.
  } as SomeObject;
}
```

**Expected Result**

When returning an object from a function (that itself has a type signature) you arguably don't need the explicit return type here as you have it in the function's type signature. It would be nice to be able to have an option to ignore this case. This is most useful when you have factory functions.

**Actual Result**

`Missing return type on function.eslint@typescript-eslint/explicit-function-return-type`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `v12.2.0` |

**Comment by bradzacher** [comment]
I've always been torn on the return case.

On the one hand in some cases it seems like it is explicit enough for a human to easily understand and follow what the type should be - like your examples of 5 line functions.

However it's not too hard for a function to grow such that it's no longer easy for a human to infer it, as the type of the return is now very disjointed from the return statement. For example if the function:
- is several hundred lines long,
- has many branches,
- has a complex union return type.

In the past I've pushed back against this. But yeah. I'm not sure...

If someone wants to do this work, I'll accept a PR.

**Comment by dman777** [comment]
This has been broken for quite awhile. But to the point of the original poster, typescript even mentions this return types of functions can be optionally left off [here](https://www.typescriptlang.org/docs/handbook/functions.html)

> TypeScript can figure the return type out by looking at the return statements, so we can also optionally leave this off in many cases.

**Comment by bodinsamuel** [comment]
> On the one hand in some cases it seems like it is explicit enough for a human to easily understand and follow what the type should be - like your examples of 5 line functions.

Hello I would +1 on this issue.
I think the rule has two usage:
- strict typing (no auto inference)
- readability 

In **terms of strict typing** the use case presented here is covered. 
Asking to put it a second time is irrelevant as it won't help, and can be error prone as you can wrongly type.
This also happen with well typed signature + object, e.g: 

```ts
/**
 * In this example Pool (from tarn package) is well typed and the signature too, 
 * the `create` prop has a return type displayed on hover but the rule still complain 
 */
import { Pool } from 'tarn';

const pool = new Pool({
  create: (cb) => { }, // Missing return type on function.
});
```

If you want to fix the `readability` aspect of this rule (aka asking to put it again even if it's already defined) maybe there is another way, but I don't see something that can not be fixed by IDE. And if you don't use IDE then I'm out of idea :D
